metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+        return dict_.kmer_forward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_backward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_backward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+  private:
+    // TODO: this is a hack since sshash::dictionary<kmer_t>::visit is always non-const
+    mutable sshash::dictionary<kmer_t> dict_;
+};
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = k_ % 2 == 0 ? k_ + 1 : k_;
+
+    if (odd_k * IDictionary::bits_per_char <= 64) {
+        dict_ = std::make_unique<Dictionary<Kmer64>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 128) {
+        dict_ = std::make_unique<Dictionary<Kmer128>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 256) {
+        dict_ = std::make_unique<Dictionary<Kmer256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / IDictionary::bits_per_char);
+        throw std::runtime_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k, Mode mode)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+    // quick fix for value of m... k/2 but odd
+    build_config.m = std::min(uint64_t((k_ + 1) / 2), sshash::constants::max_m);
+    build_config.verbose = common::get_verbose();
+    build_config.num_threads = get_num_threads();
+    if (build_config.m % 2 == 0)
+        build_config.m++;
+
+    // silence sshash construction messages when not verbose
+    if (!common::get_verbose())
+        std::cout.setstate(std::ios_base::failbit);
+
+    dict_->build(input_filename, build_config);
+    if (!common::get_verbose())
+        std::cout.clear();
+
+    num_nodes_ = dict_->size();
+}
+
+std::string DBGSSHash::file_extension() const {
+    return kExtension;
+}
+
+void DBGSSHash::add_sequence(std::string_view sequence,
+                             const std::function<void(node_index)>& on_insertion) {
+    throw std::runtime_error("adding sequences not implemented");
+}
+
+void DBGSSHash::map_to_nodes(std::string_view sequence,
+                             const std::function<void(node_index)>& callback,
+                             const std::function<bool()>& terminate) const {
+    if (mode_ != CANONICAL) {
+        map_to_nodes_sequentially(sequence, callback, terminate);
+    } else {
+        map_to_nodes_with_rc(
+                sequence, [&](node_index node, bool) { callback(node); }, terminate);
+    }
+}
+
+DBGSSHash::node_index DBGSSHash::reverse_complement(node_index node) const {
```

Comment:
How does sshash handles this? Does it have the same index for a k-mer and its reverse complement?

Reply:
In sshash, they use the same index for both orientations. They have a bool flag in the lookup_result struct that stores whether your query was in the same or the opposite orientation from the k-mer that's stored

Reply:
Since we also consider even `k` here, it might be that there are "unused" nodes for self-complement k-mers. Are we fine with that? Generally, in this case, maybe it's better to return `node + dict_->size()` anyway, so that we don't waste time on comparison of the node with its revcompl? Would it break anything if we have a "virtual" revcompl node when it is exactly equal to the initial node?

Reply:
I think there are some parts of the code that rely on the fact that nodes with different indices necessarily have different spellings. Since we don't use even k that much anyway, I'd rather make the code for that case correct rather than extra efficient.